### PR TITLE
US-823 line split with voltage level, first commit

### PR DESCRIPTION
**Pull request by laurgarn**

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 1]
```diff
@@ -929,11 +929,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The voltage level has been created")})
+    public ResponseEntity<Mono<Void>> lineSplitWithVoltageLevel(@PathVariable("studyUuid") UUID studyUuid,
+        @PathVariable("nodeUuid") UUID nodeUuid,
+        @RequestBody String createVoltageLevelAttributes) {
+        return ResponseEntity.ok().body(studyService.assertCanModifyNode(nodeUuid).then(studyService.assertComputationNotRunning(nodeUuid))
+            .then(studyService.lineSplitWithVoltageLevel(studyUuid, createVoltageLevelAttributes, ModificationType.LINE_SPLIT_WITH_VOLTAGE_LEVEL, nodeUuid, null)));
+    }
+
+    @PutMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/modifications/{modificationUuid}/line-splits")
+    @Operation(summary = "update a voltage level creation in the study network")
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/study/server/StudyController.java:950]: correct summary

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 2]
```diff
@@ -929,11 +929,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The voltage level creation has been updated.")})
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/study/server/StudyController.java:951]: correct description

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 3]
```diff
@@ -929,11 +929,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
+        return ResponseEntity.ok().body(studyService.assertComputationNotRunning(nodeUuid)
+            .then(studyService.updateEquipmentCreation(studyUuid, createVoltageLevelAttributes, ModificationType.VOLTAGE_LEVEL_CREATION, nodeUuid, modificationUuid)));
+    }
+
+    @PostMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/line-splits")
+    @Operation(summary = "create a voltage level in the study network")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The voltage level has been created")})
+    public ResponseEntity<Mono<Void>> lineSplitWithVoltageLevel(@PathVariable("studyUuid") UUID studyUuid,
+        @PathVariable("nodeUuid") UUID nodeUuid,
+        @RequestBody String createVoltageLevelAttributes) {
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/study/server/StudyController.java:944]: change parameter name

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 4]
```diff
@@ -929,11 +929,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
+            .then(studyService.lineSplitWithVoltageLevel(studyUuid, createVoltageLevelAttributes, ModificationType.LINE_SPLIT_WITH_VOLTAGE_LEVEL, nodeUuid, null)));
+    }
+
+    @PutMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/modifications/{modificationUuid}/line-splits")
+    @Operation(summary = "update a voltage level creation in the study network")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The voltage level creation has been updated.")})
+    public ResponseEntity<Mono<Void>> updateLineSplitWithVoltageLevel(@PathVariable("studyUuid") UUID studyUuid,
+        @PathVariable("modificationUuid") UUID modificationUuid,
+        @PathVariable("nodeUuid") UUID nodeUuid,
+        @RequestBody String createVoltageLevelAttributes) {
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/study/server/StudyController.java:955]: change parameter name

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 5]
```diff
@@ -2317,6 +2317,11 @@ public void testCreateVoltageLevel() {
 
     }
 
+    @Test
+    public void testLineSplitWithVoltageLevel() {
+
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2322]: add test code

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 6]
```diff
@@ -2364,6 +2369,46 @@ public void testCreateVoltageLevel() {
...
+            .build();
+        LineSplitWithVoltageLevelInfos lineSplitWoVL = new LineSplitWithVoltageLevelInfos("line3", 10.0, vl1, null, "1.A",
+            "nl1", "NewLine1", "nl2", "NewLine2");
+
+        webTestClient.post()
+            .uri("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/line-splits",
+                studyNameUserIdUuid, modificationNodeUuid)
+            .bodyValue(BodyInserters.fromValue(lineSplitWoVL))
+            .exchange()
+            .expectStatus().isOk();
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2400]: Missing checks on notification messages received

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 7]
```diff
@@ -2364,6 +2369,46 @@ public void testCreateVoltageLevel() {
...
+
+        webTestClient.put()
+            .uri("/v1/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/modifications/{modificationUuid}/line-splits",
+                studyNameUserIdUuid, modificationNodeUuid2, MODIFICATION_UUID)
+            .bodyValue(BodyInserters.fromValue(lineSplitWoVL))
+            .exchange()
+            .expectStatus().isOk();
+
+        var requests = getRequestsWithBodyDone(2);
```
Comment (FranckLecuyer) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:2409]: Missing checks on mocked rest api requests done

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

**Review by FranckLecuyer — APPROVED** [review]

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 8]
```diff
@@ -1892,5 +1894,29 @@ private void checkStudyContainsNode(UUID studyUuid, UUID nodeUuid) {
...
         }
     }
+
+    public Mono<Void> lineSplitWithVoltageLevel(UUID studyUuid, String lineSplitWithVoltageLevelAttributes,
+        ModificationType modificationType, UUID nodeUuid, UUID modificationUuid) {
+        return Mono.zip(getModificationGroupUuid(nodeUuid), getVariantId(nodeUuid)).flatMap(tuple -> {
+            UUID groupUuid = tuple.getT1();
+            String variantId = tuple.getT2();
+
+            Mono<Void> monoUpdateStatusResult = updateStatuses(studyUuid, nodeUuid);
```
Comment (FranckLecuyer) [inline src/main/java/org/gridsuite/study/server/StudyService.java:1904]: Pb here : you must pass an additional boolean parameter 'false' to updateStatuses method when you update an existing modification, and not when you create a new modification

**Review by FranckLecuyer — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/StudyService.java (hunk 8 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/StudyService.java]: Done

**Review by FranckLecuyer — APPROVED** [review]

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 9]
```diff
@@ -951,11 +951,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
+        return ResponseEntity.ok().body(studyService.assertComputationNotRunning(nodeUuid)
+            .then(studyService.updateEquipmentCreation(studyUuid, createVoltageLevelAttributes, ModificationType.VOLTAGE_LEVEL_CREATION, nodeUuid, modificationUuid)));
+    }
+
+    @PostMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/line-splits")
+    @Operation(summary = "split a line at a voltage level")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The line was split at a voltage level ok")})
+    public ResponseEntity<Mono<Void>> lineSplitWithVoltageLevel(@PathVariable("studyUuid") UUID studyUuid,
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/StudyController.java:964]: createLineSplitWithVoltageLevel ?

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 10]
```diff
@@ -951,11 +951,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
+    @PutMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/modifications/{modificationUuid}/line-splits")
+    @Operation(summary = "update a line split at a voltage level")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The line split at a voltage level has been updated.")})
+    public ResponseEntity<Mono<Void>> updateLineSplitWithVoltageLevel(@PathVariable("studyUuid") UUID studyUuid,
+        @PathVariable("modificationUuid") UUID modificationUuid,
+        @PathVariable("nodeUuid") UUID nodeUuid,
+        @RequestBody String lineSplitWithVoltageLevelAttributes) {
         return ResponseEntity.ok().body(studyService.assertComputationNotRunning(nodeUuid)
-                .then(studyService.updateEquipmentCreation(studyUuid, createVoltageLevelAttributes, ModificationType.VOLTAGE_LEVEL_CREATION, nodeUuid, modificationUuid)));
+            .then(studyService.lineSplitWithVoltageLevel(studyUuid, lineSplitWithVoltageLevelAttributes, ModificationType.LINE_SPLIT_WITH_VOLTAGE_LEVEL, nodeUuid, modificationUuid)));
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/StudyController.java:979]: updateLineSplitWithVoltageLevel ?

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 11]
```diff
@@ -951,11 +951,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
+    }
+
+    @PostMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/line-splits")
+    @Operation(summary = "split a line at a voltage level")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The line was split at a voltage level ok")})
+    public ResponseEntity<Mono<Void>> lineSplitWithVoltageLevel(@PathVariable("studyUuid") UUID studyUuid,
+        @PathVariable("nodeUuid") UUID nodeUuid,
+        @RequestBody String lineSplitWithVoltageLevelAttributes) {
+        return ResponseEntity.ok().body(studyService.assertCanModifyNode(nodeUuid).then(studyService.assertComputationNotRunning(nodeUuid))
+            .then(studyService.lineSplitWithVoltageLevel(studyUuid, lineSplitWithVoltageLevelAttributes, ModificationType.LINE_SPLIT_WITH_VOLTAGE_LEVEL, nodeUuid, null)));
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/StudyController.java:968]: Same

File: src/main/java/org/gridsuite/study/server/NetworkModificationService.java [hunk 12]
```diff
@@ -385,4 +385,52 @@ public Mono<Void> reorderModification(UUID groupUuid, UUID modificationUuid, UUI
...
 
     }
+
+    public Flux<ModificationInfos> lineSplitWithVoltageLevel(UUID studyUuid, String lineSplitWithVoltageLevelAttributes,
+        UUID groupUuid, ModificationType modificationType, String variantId, UUID modificationUuid) {
+
+        Objects.requireNonNull(studyUuid);
+        Objects.requireNonNull(lineSplitWithVoltageLevelAttributes);
+
+        if (modificationUuid == null) {
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/NetworkModificationService.java:395]: Split into 2 methods (create and update) ?

Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/NetworkModificationService.java:395]: Duplicate the code, just to mimic equipment creation hypotheses ?
I don't subscribe to this.

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 10 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/StudyController.java:979]: See below

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 11 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/StudyController.java:968]: See below

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 4 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/StudyController.java]: done

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 3 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/StudyController.java]: done

File: src/main/java/org/gridsuite/study/server/StudyController.java (hunk 2 above)
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/StudyController.java]: done

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 13]
```diff
@@ -929,11 +929,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
+        @PathVariable("modificationUuid") UUID modificationUuid,
+        @PathVariable("nodeUuid") UUID nodeUuid,
+        @RequestBody String createVoltageLevelAttributes) {
+        return ResponseEntity.ok().body(studyService.assertComputationNotRunning(nodeUuid)
+            .then(studyService.updateEquipmentCreation(studyUuid, createVoltageLevelAttributes, ModificationType.VOLTAGE_LEVEL_CREATION, nodeUuid, modificationUuid)));
+    }
+
+    @PostMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/line-splits")
+    @Operation(summary = "create a voltage level in the study network")
+    @ApiResponses(value = {@ApiResponse(responseCode = "200", description = "The voltage level has been created")})
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/StudyController.java]: done

File: src/main/java/org/gridsuite/study/server/StudyController.java [hunk 14]
```diff
@@ -929,11 +929,32 @@ public ResponseEntity<Mono<Void>> createVoltageLevel(@PathVariable("studyUuid")
...
-                                                                           @RequestBody String createVoltageLevelAttributes) {
+        @PathVariable("modificationUuid") UUID modificationUuid,
+        @PathVariable("nodeUuid") UUID nodeUuid,
+        @RequestBody String createVoltageLevelAttributes) {
+        return ResponseEntity.ok().body(studyService.assertComputationNotRunning(nodeUuid)
+            .then(studyService.updateEquipmentCreation(studyUuid, createVoltageLevelAttributes, ModificationType.VOLTAGE_LEVEL_CREATION, nodeUuid, modificationUuid)));
+    }
+
+    @PostMapping(value = "/studies/{studyUuid}/nodes/{nodeUuid}/network-modification/line-splits")
+    @Operation(summary = "create a voltage level in the study network")
```
[the remark this answers is not in the document]
Comment (laurgarn) [reply src/main/java/org/gridsuite/study/server/StudyController.java]: done

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**